### ISSUE TITLE
[Enahancement] No ssh key roll out with fido2

### ISSUE THREAD

**Issue by 53845714nF** [saltstack/salt#60183]
**Description**
When I try to roll out a public SSH key secured with fidio2 it does not work. With a normal key it works.

**Setup**
`init.sls` File: 
```
ssh-key-foobar:
  ssh_auth.present:
    - user: foobar
    - source: salt://linux/sshd/fidio.pub
```
**Steps to Reproduce the behavior**
A `.pub` file with that started with `sk-ssh-ed25519@openssh.com` does not work. I get the following error message.
```
ID: ssh-key-foobar
Function: ssh_auth.present
Result: False
Comment: No keys detected in salt://linux/sshd/fidio.pub. Is file properly formatted?
```
When I use a normal key it works (ssh-ed25519).

**Expected behavior**
The state should be rolled out normally 

**Screenshots**

**Versions Report**
<details><summary>salt --versions-report</summary>

```
Salt Version:
          Salt: 3003

Dependency Versions:
          cffi: Not Installed
      cherrypy: Not Installed
      dateutil: 2.7.3
     docker-py: Not Installed
         gitdb: 2.0.5
     gitpython: 2.1.11
        Jinja2: 2.10
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 0.5.6
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: Not Installed
      pycrypto: Not Installed
  pycryptodome: 3.6.1
        pygit2: Not Installed
        Python: 3.7.3 (default, Jan 22 2021, 20:04:44)
  python-gnupg: Not Installed
        PyYAML: 3.13
         PyZMQ: 17.1.2
         smmap: 2.0.5
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.1

System Versions:
          dist: debian 10 buster
        locale: utf-8
       machine: x86_64
       release: 5.4.106-1-pve
        system: Linux
       version: Debian GNU/Linux 10 buster

```
</details>

**Additional context**
None

**Comment by dmurphy18** [comment]
@53845714nF Not sure if Salt supports fido2, checking but not finding support for it in the code.

**Comment by dmurphy18** [comment]
@53845714nF Currently Salt does not support fido2 from https://fidoalliance.org/fido-authentication.  You could turn this issue into a Feature Request (Label Feature) if desired, or Salt would be more than happy to accept contributions adding fido2.

Please consider closing this issue or turning it into a Feature Request, given that fido2 is currently unsupported in Salt

**Comment by 53845714nF** [comment]
@dmurphy18 
It is not really a Feature request. It is much more a bug in rolling out SSH public keys. As of SSH 8.2+ fidio2 is supported. 
The fidio2 support does not really concern Salt.

**Comment by dmurphy18** [comment]
@53845714nF 
```
As of SSH 8.2+ fidio2 is supported.
```
Can you point me to where there is a statement in Salt Project doc's that states fido2 is supported - it needs correction if it is there.

If the statement is because SSH 8.2+ supports fido2, then there is an issue in the version SSH that Salt is supporting.  Looking for more context, since I have not been following SSH lately, as focused on other UNIX platforms.

**Comment by 53845714nF** [comment]
I expect to be able to distribute my ssh keys no matter what format they have.
And the key is in a valid ssh format. That is the point.

Where in the documentation are the formats that are not supported?

> Could it be the OS doesn't support fido2 on Debian 10 yet ?

The normal Debian repositories still have an old ssh, it has to be installed from the backports.

Here is a simple state that does it:
```
buster-backports:
  pkgrepo.managed:
    - humanname: Buster backports repository
    - name: deb http://deb.debian.org/debian buster-backports main
    - dist: buster-backports
    - file: /etc/apt/sources.list.d/buster-backports.list

ssh:
  pkg.installed:
    - fromrepo: buster-backports
    - version: 1:8.4p1-2~bpo10+1
  ```

But it doesn't really matter. With some operating systems like Arch and Alpine it is already the standard.

**Comment by dmurphy18** [comment]
@53845714nF You reported Debian 10, not Arch and Alpine. Can only test on what is reported.
Will look at a version of Debian 10 with backports enabled, but NOTE, not all users of Debian enable backports, this has been an issue in the past where organizations using Debian did not use backports and could not simply upgrade to receive the fix, thus resulting in work-arounds.

Lastly, valid ssh format, typically older code cannot parse newer formats or additions, and depending on the code can ignore or throw errors.  In this age, code no longer tends to follow Jon Postel's benign skip of stuff not recognized.

Will dig into this further.

**Comment by 53845714nF** [comment]
@dmurphy18 Yes, my salt master is a Debian 10 with the SSH version OpenSSH_8.4p1. 
The other operating systems referred to my Minions, they can use different operating systems. 
I don't think my Salt master is particularly old.


>  typically older code cannot parse newer formats or additions,

This is the reason for my Issue.

**Comment by dmurphy18** [comment]
@53845714nF Ubuntu 20.04.2 has ssh v8.2, so this definitely has to be fixed, just ordered a YubiKey 5 NFC to ensure all of this functions as expected.  So have to enhance support of ssh to ensure fido2 is handled correctly since now in stable OS with LTS.

**Comment by 53845714nF** [comment]
Great :+1:

**Comment by 53845714nF** [comment]
Could you reproduce my error?

**Comment by dmurphy18** [comment]
@53845714nF Would need to generate the relevant pub key file, but given the ssh support in Ubuntu 20.04, need to update to allow for Fido2, so was waiting for key to arrive.  Received my YubiKey over the Memorial Day weekend (US holiday), so will be looking at this soon, have some other fixes need to complete first but it is on my hot list.

**Comment by 53845714nF** [comment]
Hello, sounds good. I found a similar bug also in ProxmoxVE, there only one regex had to be adjusted. 
Do you know in which file the ssh key is checked? Maybe I can have a look at it and create a pull request.

**Comment by 53845714nF** [comment]
Is there anything new? 
Debian 11 will be released next month, with the new ssh version.

**Comment by dmurphy18** [comment]
@53845714nF Had to put work on hold while building native minions for recent Salt release 3003.1.  Once I have Salt 3003.1 for Arista, Juniper Solaris 10 and 11, AIX shall be back onto it.

Ubuntu 20.04 already has support, so urgency is real.

**Comment by 53845714nF** [comment]
I have seen commit 88658aab2f8ba1a636efa384a98002dd86d81e48.
Does this fix my issue?

Best regards

**Comment by dmurphy18** [comment]
I don't know, have been dragged through various hot-flaming issues and still have this issue on my list to address.
Got my yubico key and about the only thing I have been able to do is open the package.

Have called out to management that we need to address this, so it is on the schedule for Phosphorous

**Comment by 53845714nF** [comment]
I think @Ch3LL and @MEschenbacher have already improved this. I'm looking forward to the next release.

**Comment by dmurphy18** [comment]
@53845714nF Any chance of checking out the master branch and testing on a throw-away VM to see if commit https://github.com/saltstack/salt/commit/88658aab2f8ba1a636efa384a98002dd86d81e48 resolves your issue ?.  If so, could close this, or leave it around with a verify for the Phosphorous release

**Comment by 53845714nF** [comment]
I have to see what I can do.
At the moment I have little time for such things.

**Comment by dmurphy18** [comment]
@53845714nF 

Going to close this as I believe it is now resolved on the master branch ( commit https://github.com/saltstack/salt/commit/88658aab2f8ba1a636efa384a98002dd86d81e48 ).
Using Master and Minion with master branch as of 2022-02-13 got the following result, using a Yubikey key generated pub file

```
root@c7mstr:/srv/salt# salt tc7 state.sls fido2
tc7:
----------
          ID: ssh_key_for_bar
    Function: ssh_auth.present
      Result: True
     Comment: The authorized host key ssh_key_for_bar for user david was added
     Started: 17:00:53.728389
    Duration: 47.529 ms
     Changes:   
              ----------
              ssh_key_for_bar:
                  New

Summary for tc7
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  47.529 ms
root@c7mstr:/srv/salt# 
```
where fido2.sls is 
```root@c7mstr:/srv/salt# cat fido2.sls 
ssh_key_for_bar:
  ssh_auth.present:
    - user: david
    - source: salt://linux/sshd/fidio.pub

#    - user: foobar
root@c7mstr:/srv/salt# 
```
and fidio.pub contains the following:
```
sk-ecdsa-sha2-nistp256@openssh.com AAAAAA ------------------ Ec3NoOg== testhost-14-02-2022-physical_yubikey_number
```
Was using Centos 7.9 for Master and Minion, key generated on Ubuntu 21.10 machine.

And again with ed25519_sk this time
```
root@c7mstr:/srv/salt# cat fido2_ed.sls
ssh_key_for_bar:
  ssh_auth.present:
    - user: david
    - source: salt://linux/sshd/fidio_ed25519sk.pub
root@c7mstr:/srv/salt# salt tc7 state.show_sls fido2_ed
tc7:
    ----------
    ssh_key_for_bar:
        ----------
        __env__:
            base
        __sls__:
            fido2_ed
        ssh_auth:
            |_
              ----------
              user:
                  david
            |_
              ----------
              source:
                  salt://linux/sshd/fidio_ed25519sk.pub
            - present
            |_
              ----------
              order:
                  10000
root@c7mstr:/srv/salt# salt tc7 state.sls fido2_ed
tc7:
----------
          ID: ssh_key_for_bar
    Function: ssh_auth.present
      Result: True
     Comment: The authorized host key ssh_key_for_bar for user david was added
     Started: 17:10:51.114992
    Duration: 37.268 ms
     Changes:   
              ----------
              ssh_key_for_bar:
                  New

Summary for tc7
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  37.268 ms
root@c7mstr:/srv/salt# 
```
where fidio_ed25519sk.pub
```
sk-ssh-ed25519@openssh.com AAAAGBBBBBBBBBBBBB--------------------------3YAAAABHNzaDo= testhost-14-02-2022-physical_yubikey_number
```

Note: the pub key files middle contents modified to be non-functional, never can be too safe.

Please re-open the issue if there is some other information which is necessary to resolve it.

The changes should be available with the next major release of Salt, Phosphorous
